crates/wasi/src/preview2/preview1/mod.rs

```diff
@@ -1,32 +1,289 @@
+        let directories = preopens
+            .get_directories()
+            .await
+            .context("failed to call `get-directories`")
+            .map_err(types::Error::trap)?;
+
+        let mut descriptors = Self::default();
+        descriptors.push(Descriptor::Stdin(stdin))?;
+        descriptors.push(Descriptor::Stdout(stdout))?;
+        descriptors.push(Descriptor::Stderr(stderr))?;
+        for dir in directories {
+            descriptors.push(Descriptor::PreopenDirectory(dir))?;
+        }
+        Ok(descriptors)
+    }
+
+    fn unused(&self) -> ErrnoResult<u32> {
+        match self.last_key_value() {
+            Some((fd, _)) => {
+                if let Some(fd) = fd.checked_add(1) {
+                    return Ok(fd);
+                }
+                if self.len() == u32::MAX as usize {
+                    return Err(types::Errno::Loop);
+                }
+                // TODO: Optimize
+                Ok((0..u32::MAX)
+                    .rev()
+                    .find(|fd| !self.contains_key(fd))
+                    .expect("failed to find an unused file descriptor"))
+            }
+            None => Ok(0),
+        }
+    }
+
+    fn remove(&mut self, fd: types::Fd) -> Option<Descriptor> {
+        let fd = fd.into();
+        let desc = self.used.remove(&fd)?;
+        self.free.push(fd);
+        Some(desc)
+    }
+
+    fn push(&mut self, desc: Descriptor) -> ErrnoResult<u32> {
+        let fd = if let Some(fd) = self.free.pop() {
+            fd
+        } else {
+            self.unused()?
+        };
+        assert!(self.insert(fd, desc).is_none());
+        Ok(fd)
+    }
+
+    fn push_file(&mut self, file: File) -> ErrnoResult<u32> {
+        self.push(Descriptor::File(file))
+    }
+}
 
-pub struct WasiPreview1Adapter {/* all members private and only used inside this module. also, this struct should be Send. */}
 impl WasiPreview1Adapter {
-    // This should be the only public interface of this struct. It should take
-    // no parameters: anything it needs from the preview 2 implementation
-    // should be retrieved lazily.
     pub fn new() -> Self {
-        Self {}
+        Self::default()
     }
 }
 
 // Any context that needs to support preview 1 will impl this trait. They can
 // construct the needed member with WasiPreview1Adapter::new().
-pub trait WasiPreview1View: Send {
+pub trait WasiPreview1View: Send + Sync + WasiView {
     fn adapter(&self) -> &WasiPreview1Adapter;
     fn adapter_mut(&mut self) -> &mut WasiPreview1Adapter;
 }
 
-// This becomes the only way to add preview 1 support to a wasmtime (module)
-// Linker:
+struct DescriptorsRef<'a, T: WasiPreview1View + ?Sized> {
+    view: &'a mut T,
+    descriptors: Cell<Descriptors>,
+}
+
+impl<T: WasiPreview1View + ?Sized> Drop for DescriptorsRef<'_, T> {
+    fn drop(&mut self) {
+        let descriptors = self.descriptors.take();
+        self.view.adapter_mut().descriptors = Some(descriptors);
+    }
+}
```

Comment:
I wasn't expecting this sort of machinery to be required - we can't get away with implementing the DescriptorsRef methods on WasiPreview1Ctx? I can't intuit where the problem is in a more straightforward implementation, so if this is required, lets leave a comment on the struct definition explaining why.

Reply:
I'll add a comment, but the biggest reasons why this is required are:
1. By far the biggest issue is `bindgen` generating host methods with `&mut self` receivers, preventing e.g. simultaneous lazy initialization of the descriptor table and calling a preview2 method using any data borrowed from that table, even looking up something like `table().is_file` is problematic
2. `WasiPreview1Adapter` (assuming that's what you meant by WasiPreview1Ctx) does not have access to preview2 functionality and/or the underlying table, so it cannot initialize itself using something like https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html . Lazily initializing the descriptor map requires access to both the `WasiPreview1Adapter` and `wasi::preopens::Host`, so the only place where this functionality could exist is in an extension trait (`WasiPreview1ViewExt`). Note that from within this trait, we can never return anything borrowed directly, because we'd first need to call `adapter_mut()` to get access to the state (which we may need to lazily initialize). And even if we had a way to return a borrowed descriptor table directly, it would not be possible to call any other method on `self` due to 1., since that would cause a double mutable borrow, which is not allowed.

In short, an alternative solution would be storing `Option<Arc<Mutex<Descriptors>>>` in the adapter and returning `Arc<Mutex<Descriptors>>` in the `WasiPreview1AdapterExt::descriptors(&mut self)` method, which IMO is way more awkward to work with and simply redundant complexity.
Overall, the `DescriptorsRef` is just a workaround for `bindgen` design drawbacks. I believe this could be simplified dramatically if host methods would just take `&self`

Reply:
That makes sense. Thank you.

We should look into adding a bindgen option to enable &self receivers in the future, and then we may be able to unwind a bit of this complexity. For now, this is a fine implementation once it has a comment.

Reply:
I've added a few comments and renamed the struct to `Transaction` so it's a bit more obvious what it actually is